[lldb] Introduce an always-on system log category/channel

JDevlieghere opened the pull request:
Add an "always on" log category and channel. Unlike other, existing log channels, it is not exposed to users. The channel is meant to be used sparsely and deliberately for logging high-value information to the system log.

We have a similar concept in the downstream Swift fork and this has proven to be extremely valuable. This is especially true on macOS where system log messages are automatically captured as part of a sysdiagnose.

This patch revives https://reviews.llvm.org/D135621 although the purpose is slightly different (logging to the system log rather than to a ring buffer dumped as part of the diagnostics). This patch addresses all the remaining oustanding comments.

jasonmolenda commented on lldb/include/lldb/Utility/LLDBLog.h:
```diff
@@ -49,6 +49,7 @@ enum class LLDBLog : Log::MaskType {
   Watchpoints = Log::ChannelFlag<30>,
   OnDemand = Log::ChannelFlag<31>,
   Source = Log::ChannelFlag<32>,
+  AlwaysOn = Log::ChannelFlag<32>,
```
does this number need to be unique?

JDevlieghere replied:
Yes, definitely. But this code shouldn't have been part of this PR (I created a new category like I did on Phab). Let me clean this up.

DavidSpickett commented on lldb/include/lldb/Host/Host.h:
```diff
@@ -31,6 +31,22 @@ class ProcessInstanceInfo;
 class ProcessInstanceInfoMatch;
 typedef std::vector<ProcessInstanceInfo> ProcessInstanceInfoList;
 
+// Always on system log category and channel.
```
So is the log naming `system system` or for now is it just `system` and we might add sub-channels later?

adrian-prantl replied:
I think we should have a comment that explains what this channel is supposed to be used for?

DavidSpickett commented:
> Unlike other, existing log channels, it is not exposed to users.

This is presumably because:
1. It contains very low level messages they likely won't be interested in.
2. If it weren't hidden, it would be printing into their sessions all the time.

Users can still list it, I see that from the test, but that's fine we don't need to lie about its existence.

If someone suspects that a relevant message might be in this channel how do they look at it? Do they crash lldb / use a command to dump an archive of logs?

> The channel is meant to be used sparsely and deliberately for logging high-value information to the system log.

What is an example of that? So that folks (aka me) have an idea what might go here for other OSs. Is it stuff like this process was killed for this reason, this system resource had to be cleared, that sort of thing?

DavidSpickett commented:
>We have a similar concept in the downstream Swift fork and this has proven to be extremely valuable. This is especially true on macOS where system log messages are automatically captured as part of a sysdiagnose.

Does this mean someone has already been putting non-MacOS messages in there in the Swift fork? Would be interested to know what those are.

labath commented:
My first thought when seeing (before opening the phabricator patch) this was "wouldn't this be better off as a separate API". After opening the link, I saw that this was the last comment on it as well. So, what has changed since then?

My think is that, our logging API is a fairly complex piece of machinery, and I think this is largely due to the fact that we're allowing the user to customize it (change the target, verbosity, etc.). For the this log channel, you basically take away all of that, so what's the point in going through all of that? Why not just have a single `SystemLog` function that does what you want?

JDevlieghere commented:
> > Unlike other, existing log channels, it is not exposed to users.
> 
> This is presumably because:
> 
> 1. It contains very low level messages they likely won't be interested in.
> 2. If it weren't hidden, it would be printing into their sessions all the time.
> 
> Users can still list it, I see that from the test, but that's fine we don't need to lie about its existence.

Actually the test is testing that it *cannot* (`CHECK-NOT`) be listed. The goal is to make the fact that it's a log channel and implementation detail, as if we had used a dedicated API as Pavel suggested on Phab and here. 

> If someone suspects that a relevant message might be in this channel how do they look at it? Do they crash lldb / use a command to dump an archive of logs?

That depends on the system they're on: on macOS that would be by looking at Console.app. On Linux it's a file in `/var/log`. I wouldn't expect someone that isn't an LLDB developer to ever have to look at these logs though. 

> What is an example of that? So that folks (aka me) have an idea what might go here for other OSs. Is it stuff like this process was killed for this reason, this system resource had to be cleared, that sort of thing?

A really trivial example is logging the lldb version in the system log, which is something we do already do on the Swift fork. At Apple we have different toolchains with different versions of LLDB and surprisingly often a bug report will state the wrong version of LLDB. At this point this has probably saved me hours of wasted time. 

> Does this mean someone has already been putting non-MacOS messages in there in the Swift fork? Would be interested to know what those are.

Not quite, the Swift fork uses its own [category & channel](https://github.com/swiftlang/llvm-project/blob/next/lldb/source/Plugins/Language/Swift/LogChannelSwift.cpp). Unlike the always-on channel I'm proposing here, the Swift one is meant to be user-visible. Users can generate the log file with a command (`swift-healthcheck`) and are expected to be able to diagnose things such as project misconfigurations. 

If you look at the implementation, you can see that it's using a T-loghandler that writes both to a buffer and to the system log. 

> My first thought when seeing (before opening the phabricator patch) this was "wouldn't this be better off as a separate API". After opening the link, I saw that this was the last comment on it as well. So, what has changed since then?

I wouldn't say anything has changed, but here's why I'm proposing the log channel approach again: 

 - We've had a very positive experience with the Swift log category. I'm the screener for the LLDB component so that's something I've witnessed (and benefited from) first hand. I'm sure Dave and Adrian can share a bunch of success stories as they deal with the actual details. Of course that says more about the concept than the implementation. 
 - Everyone (else) I talked to liked the idea of reusing the logging system because that's what we already use and are familiar with. Promoting a log message to be part of this log becomes as easy as adding a `GetLog` and using the same macro. 
 - I expect that in a bunch of places, you would like the system log message to go both to the system log and to a user-visible log. If something egregious goes wrong related to types, you probably want to log it to the types log as well. With a separate API, you'll end up doing something like create the log message, then pass it to `LLDB_LOG` and then calling `Host::SystemLog`. We have a few examples where we do the same dance for Progress events and Signposts (`LLDB_SCOPED_TIMER`) which I'm also trying to solve in https://github.com/llvm/llvm-project/pull/108498. 

Ultimately I care more about the functionality than the implementation so I'm not opposed to the separate API, but I also want to encourage adoption across LLDB.

labath commented:
> * Everyone (else) I talked to liked the idea of reusing the logging system because that's what we already use and are familiar with. Promoting a log message to be part of this log becomes as easy as adding a `GetLog` and using the same macro.

I'm all for code reuse, and reusing some part of the logging machinery for this makes sense to me. The thing I'm not sure about is whether this is plugging into the machinery at the right level. The most irritating part is the `internal` flag -- like, we do all the motions to place the channel into the global log channel map (which is the source of truth for all user-facing log commands), but then set this flag which basically says "pretend this channel isn't here". 

I think this would look better if the flag wasn't there. Like, what if we had a separate way to create/register one of these "system" channels -- one which does not involve putting it in a list with all the other channels. Maybe the `Initialize` function could be something like:
```
Log g_system(g_system_channel);
g_system->Enable(...);
```
(The `Enable` function is currently private, but I think it would make sense to make it public, especially since it's possible (like you're effectively doing now) to bypass this restriction by calling the static string-based EnableLogChannel function. Alternatively, we could make the existence of this channel a feature of the generic log machinery (I expect we'll never want to have more than one of these), and then we can do this work from within `Log::Initialize`)

JDevlieghere committed:
[lldb] Introduce an always-on system log category/channel

Add an "always on" log category and channel. Unlike other, existing log
channels, it is not exposed to users. The channel is meant to be used
sparsely and deliberately for logging high-value information to the
system log.

We have a similar concept in the downstream Swift fork and this has
proven to be extremely valuable. This is especially true on macOS where
system log messages are automatically captured as part of a sysdiagnose.

This patch revives https://reviews.llvm.org/D135621 although the purpose
is slightly different (logging to the system log rather than to a ring
buffer dumped as part of the diagnostics). This patch addresses all the
remaining outstanding comments.

JDevlieghere commented:
Rebased + addressed @labath and @adrian-prantl's feedback.

labath commented on lldb/source/API/SystemInitializerFull.cpp:
```diff
@@ -83,6 +85,9 @@ llvm::Error SystemInitializerFull::Initialize() {
   // Use the Debugger's LLDBAssert callback.
   SetLLDBAssertCallback(Debugger::AssertCallback);
 
+  // Log the LLDB version to the system log.
```
I don't think this tells us anything new.

labath commented on lldb/include/lldb/Utility/Log.h:
```diff
@@ -272,6 +272,12 @@ class Log final {
 
+  void Enable(const std::shared_ptr<LogHandler> &handler_sp,
+              uint32_t options = 0,
+              MaskType flags = std::numeric_limits<Log::MaskType>::max());
```
The log channels have a concept of a default set of categories, so if we're going to have a default here, it better be it. That could be achieved by making this an `optional<MaskType>` (and then you can also simplify the code in `EnableLogChannel`).

labath commented on lldb/source/API/SystemInitializerFull.cpp:
```diff
@@ -20,6 +20,8 @@
 #include "llvm/Support/CommandLine.h"
 #include "llvm/Support/TargetSelect.h"
 
+#include "lldb/Version/Version.h"
```
Group this with the other lldb includes?

antmox commented:
Hi, could this patch be the cause of lldb-aarch64-windows bot failure ?
https://lab.llvm.org/buildbot/#/builders/141/builds/3100
